src/auth.py

```diff
@@ -0,0 +1,143 @@
+import base64
+import hashlib
+import requests
+import random
+import string
+import datetime
+import urllib
+
+from requests_toolbelt.multipart.encoder import MultipartEncoder
+
+from bb2 import Bb2
+
+BB2_AUTH_URL = "{}/v{}/o/authorize"
+BB2_TOKEN_URL = "{}/v{}/o/token/"
+
+
+class AuthRequest:
+
+    def __init__(self, bb: Bb2):
+        self.bb = bb
+        self.auth_base_url = BB2_AUTH_URL.format(bb.get_config()['base_url'], bb.get_config()['version'])
+        self.auth_token_url = BB2_TOKEN_URL.format(bb.get_config()['base_url'], bb.get_config()['version'])
+        self.auth_data = self._generate_authdata()
+        self.auth_url = self._generate_authorize_url()
+        self.auth_token = None
+
+    def get_authorize_url(self):
+        return self.auth_url
+
+    def authorize_callback(self, code, state):
+        if code is None:
+            raise ValueError("Authorization code missing.")
+
+        if state is None:
+            raise ValueError("Callback parameter 'state' missing.")
+
+        if state != self.auth_data['state']:
+            raise ValueError("Provided callback state does not match.")
+
+        self.auth_token = AuthorizationToken(self._get_access_token(code))
+
+        return self.auth_token
+
+    def access_token_expired(self):
+        return self.auth_token.access_token_expired()
+
+    def refresh_access_token(self):
+
+        if self.auth_token.refresh_token is None:
+            raise ValueError("Refresh token not available when calling refresh_access_token().")
+
+        params = {
+            'client_id': self.bb.get_config()['client_id'],
+            'grant_type': 'refresh_token',
+            'refresh_token': self.auth_token.refresh_token
+        }
+
+        token_response = requests.post(url=self.auth_token_url,
+                                       params=params,
+                                       auth=(self.bb.get_config().get('client_id'),
+                                             self.bb.get_config().get('client_secret')))
+
+        if token_response.status_code == 200:
+            self.auth_token = AuthorizationToken(token_response.json())
+        else:
+            raise Exception("Failed to refresh access token, status_code: {}, error: {}".format(
+                token_response.status_code, token_response.content))
+
+        return self.auth_token
+
+    def _generate_authorize_url(self):
+        params = {'client_id' : self.bb.get_config()['client_id'],
+                  'redirect_uri' : self.bb.get_config()['callback_url'],
+                  'state' : self.auth_data['state'],
+                  'response_type' : 'code'}
+        
+        if self.bb.get_config().get('pkce'):
```

Comment:
For our SDK's we should only allow/force PCKE and state usage. We shouldn't allow the option to not use PKCE.